File: 121. Best Time to Buy and Sell Stock/121.kt

import kotlin.math.*



class Solution{
	fun maxProfit(prices: IntArray): Int{

	val n= prices.size
	var mini= prices[0]
	var profit =0

	for(i in 0 until n){
	var cost = prices[i]-mini
	profit= max(profit,cost)
	mini=min(mini,prices[i])
	}

	return profit

	}
}

fun main(){
	val nStr= readlnOrNull() ?: return

	val n =nStr.trim().toInt()

	 if(n<1 || n>100000){
	 println("Invalid n")
	 return
	 }

	 val prices = readln().trim().split("\\s+".toRegex()).map { it.toInt()}.toIntArray()

	 for(prices in prices){
	 if(prices <0 || prices > 10000){
	 print("Invalid Price")
	 return
	 }
	 }

	 val obj = Solution()
	 val ans = obj.maxProfit(prices)

	 print(ans)

}

File: 121. Best Time to Buy and Sell Stock/121.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution{
public:
int maxProfit(vector<int>& prices){

	int n= prices.size();
	int mini=prices[0];
	int profit = 0;
	
	for(int i=0;i<n;i++){
		int cost=prices[i]-mini;
		 profit=max(profit,cost);
		mini=min(prices[i],mini);
	}
	return profit;
	 

	
}
};

int main(){
	int n;
	cin>>n;

	if(n <1 || n>100000){
		cout<<"Invalid n\n";
		return 0;
	}

	vector<int> prices(n);
	for(int i=0;i<n;i++){
		cin>>prices[i];

		if(prices[i]<0 || prices[i] >10000){
			cout<<"Invalid Price";
			return 0;
		}
	}

	Solution obj;


	int ans = obj.maxProfit(prices);

	cout<<ans;
	return 0;
	



}
